### ISSUE TITLE
Move to async network requests

### ISSUE THREAD

**Issue by futureshape** [stravalib/stravalib#349]
Currently network requests in https://github.com/stravalib/stravalib/blob/master/stravalib/protocol.py are done using the `requests` library whihch blocks while data is being downloaded. 

This isn't a problem when the libray is being used to bulk-download data but is more of an issue when I've attempted to use stravalib to build interactive applications and want to keep the rest of the UI responsive (e.g. showing a progress display, having a way to cancel a long request and move somewhere else).

It looks like there's a library that supports async requests and is pretty much a drop-in replacement for `requests` https://www.python-httpx.org/compatibility/

I'm happy to give it a go in a branch and make a PR, but I'm opening this first as requested in the contribution guideline, and to get feedback from the rest of the maintainers. Does anyone think it's a bad idea to make this library use async network requests? Anything that could trip us up while trying to do it?

**Comment by jsamoocha** [comment]
Allowing async requests would be a great improvement; thanks for bringing this up!

> Anything that could trip us up while trying to do it?

From the top of my mind (I may be overlooking stuff):

- From an API perspective, what would be the behavior and return type of the stravalib `Client` methods? E.g., what would `get_activity()` return? An actual `Activity` object or a future? Will this client method itself need to be declared as `async`?
- Would we need a separate `ASyncClient` API?
- There's a `BatchedResultsIterator` for dealing with paginated responses. When using it in an async fashion, do we need to preserve the order of the individual responses/pages?
- The current integration test suite is written using the `responses` library for mocking HTTP requests. This would need to be rewritten using a mocking library compatible with `httpx` or `aiohttp`.

Please let us know what you think about the above considerations.

If you need a short-term workaround for your particular problem, you may consider using "plain" multithreading, see e.g. https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor.

**Comment by futureshape** [comment]
Thanks Jonatan - all good questions, and I don't have an answer yet as I'm new to this part of Python. But I'm willing to give it a try and do some research to see how other libraries have done similar things.

You're right, for now I can work around the issue as you suggested but it would be nicer to make my code cleaner the future 😄

**Comment by jsamoocha** [comment]
It would be great if you have the time and resources to research this, as it is not entirely trivial. I'm happy to help if you have questions about the stravalib code base.

**Comment by yihong0618** [comment]
If we change to `async` do we need to still support  `sync` mode?

**Comment by jsamoocha** [comment]
> If we change to `async` do we need to still support `sync` mode?

I guess so. I wouldn't want to force users to deal with the added complexity of async requests, they should be able to choose.

**Comment by lwasser** [comment]
if we did try to support both, to hans' point above, the implementation
(our api) would potentially become more complex? As would testing.

Unless we wrapped some functionality around a plain multithread solution as
jonatan suggested above. Is that an option too or is that approach inferior
?

**Comment by hozn** [comment]
Yeah, I haven't seen an elegant way to support both sync and async. (I've seen libraries that support multiple asyncio frameworks, but even that can get fairly complex.)  Probably the path to supporting both means two different client classes with different underpinnings.  Unfortunately if models are set up to support lazily loading things over the request, they also would need to be specific.  That sounds tangled.  (This, incidentally, is similar to why SqlAlchemy has such challenges w/ asyncio, despite the db drivers supporting asyncio for awhile -- the lazy loading of attrs means that the entire ORM framework needs to pick to be either sync or async and that's a tall order.)

Obviously this is not my project anymore (I'm eternally grateful to you all for breathing new life into this and for taking this on!) , but personally, I don't think there's any clear mandate to support asyncio: I think it's easy enough to fix any performance issues by using multiple threads.  And folks that want to use in context of asyncio have options like `asyncio.to_thread()` or `loop.run_in_executor()`.  I don't see stravalib as supporting the true use-case of asyncio where you might have thousands of concurrent requests (thousands of pthreads would be expensive!); Strava's api rate limits would typically preclude that usage pattern.  I think the reason *to* move to asyncio would just be if the developers felt strongly opinionated that it was the right way to go.  (I think, though, in the context of data science applications, jupyter notebooks, etc. asyncio is definitely a less popular paradigm.)

**Comment by jsamoocha** [comment]
> Unless we wrapped some functionality around a plain multithread solution as jonatan suggested above

I wasn't entirely clear, but my suggestion was to use multithreading on the user application side, not as part of stravalib. E.g., when you have a bunch of activity ids, and you want to fetch their streams as fast as possible (within your rate limits), you could use something like 

```python
with ThreadPoolExecutor(max_workers=10) as executor:
    streams = executor.map(client.get_activity_streams, ids)
```

which would prevent each call of `get_activity_streams()` from waiting on the completion of the previous call. Instead of `map()`, you could also use `submit()` to immediately return control without waiting on the result.
